libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,383 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/ioutil"
+	"github.com/keybase/kbfs/kbfsblock"
+	"github.com/keybase/kbfs/kbfscrypto"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/pkg/errors"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"golang.org/x/net/context"
+)
+
+const (
+	// 10 GB maximum storage by default
+	defaultDiskBlockCacheMaxBytes uint64 = 10 * (1 << 30)
+	evictionConsiderationFactor   uint   = 3
+	blockDbFilename               string = "diskCacheBlocks.leveldb"
+	lruDbFilename                 string = "diskCacheLRU.leveldb"
+	versionFilename               string = "version"
+	initialDiskCacheVersion       uint64 = 1
+	currentDiskCacheVersion       uint64 = initialDiskCacheVersion
+)
+
+// diskBlockCacheEntry packages an encoded block and serverHalf into one data
+// structure, allowing us to encode it as one set of bytes.
+type diskBlockCacheEntry struct {
+	Buf        []byte
+	ServerHalf kbfscrypto.BlockCryptKeyServerHalf
+}
+
+// diskBlockCacheConfig specifies the interfaces that a DiskBlockCacheStandard
+// needs to perform its functions. This adheres to the standard libkbfs Config
+// API.
+type diskBlockCacheConfig interface {
+	codecGetter
+	logMaker
+	clockGetter
+}
+
+// DiskBlockCacheStandard is the standard implementation for DiskBlockCache.
+type DiskBlockCacheStandard struct {
+	config   diskBlockCacheConfig
+	maxBytes uint64
+	log      logger.Logger
+	// This protects the disk caches from being shutdown while they're being
+	// accessed.
+	lock    sync.RWMutex
+	blockDb *leveldb.DB
+	lruDb   *leveldb.DB
+}
+
+var _ DiskBlockCache = (*DiskBlockCacheStandard)(nil)
+
+// openLevelDB opens or recovers a leveldb.DB with a passed-in storage.Storage
+// as its underlying storage layer.
+func openLevelDB(stor storage.Storage) (db *leveldb.DB, err error) {
+	db, err = leveldb.Open(stor, leveldbOptions)
```

Comment:
Might be for a future PR, but would be nice to wrap all errors coming from leveldb with `errors.WithStack` (and in general all the errors this file generates that aren't already wrapped)